assets/app/scripts/services/data.js

```diff
@@ -360,6 +360,122 @@ angular.module('openshiftConsole')
+      if(!_.isFunction(fn)) {
+        return;
+      }
+      var id = _.uniqueId(identifier);
+      openQueue[id] = fn;
+      return id;
+    },
+    message: function(fn) {
+      if(!_.isFunction(fn)) {
+        return;
+      }
+      var id = _.uniqueId(identifier);
+      messageQueue[id] = fn;
+      return id;
+    },
+    close: function(fn) {
+      if(!_.isFunction(fn)) {
+        return;
+      }
+      var id = _.uniqueId(identifier);
+      closeQueue[id] = fn;
+      return id;
+    },
+    error: function(fn) {
+      if(!_.isFunction(fn)) {
+        return;
+      }
+      var id = _.uniqueId(identifier);
+      errorQueue[id] = fn;
+      return id;
+    },
+    // can remove any callback from open, message, close or error
+    remove: function(id) {
+      if (openQueue[id]) { delete openQueue[id]; }
+      if (messageQueue[id]) { delete messageQueue[id]; }
+      if (closeQueue[id]) { delete closeQueue[id]; }
+      if (errorQueue[id]) { delete errorQueue[id]; }
+    },
+    stream: function() {
+      return getNamespace(kind, context, {})
+                .then(function(params) {
+                  return  $ws({
+                            url: urlForResource(kind, name, null, context, true, _.extend(params, {follow: true})),
+                            onopen: function(evt) {
+                              _.each(openQueue, function(fn) {
+                                fn(evt);
+                              });
+                            },
+                            onmessage: function(evt) {
+                              if(!_.isString(evt.data)) {
+                                Logger.log('log stream response is not a string', event.data);
+                                return;
+                              }
+                              _.each(messageQueue, function(fn) {
+                                if(isBinary) {
+                                  fn(evt.data);
+                                } else {
+                                  fn(b64_to_utf8(evt.data), evt.data);
+                                }
+                              });
+                            },
+                            onclose: function(evt) {
+                              _.each(closeQueue, function(fn) {
+                                fn(evt);
+                              });
+                            },
+                            onerror: function(evt) {
+                              _.each(errorQueue, function(fn) {
+                                fn(evt);
+                              });
+                            },
+                            protocols: protocols
+                          }).then(function(ws) {
+                            Logger.log("Streaming pod log", ws);
+                            return ws;
+                          });
+                });
+
+    }
+  };
+};
+
+// handle = the websocket (for now)
+DataService.prototype.closeStream = function(handle) {
```

Comment:
so handle is what would be returned from createStream().stream()?  doesn't feel right
As someone consuming this I would expect it to behave in one of these two ways:
```
var myStream = ds.createStream()
// do stuff with myStream
ds.closeStream(myStream);
```

or

```
var myStream = ds.createStream()
// do stuff with myStream
myStream.closeStream(myStream);
```

I'm leaning towards the second one.  And then renaming stream to openStream or startStream or something.